dotnet/src/Connectors/VectorData.Abstractions/ConnectorSupport/VectorStoreRecordModel.cs

```diff
@@ -0,0 +1,240 @@
+
+    /// <summary>
+    /// The key properties of the record.
+    /// </summary>
+    public IReadOnlyList<VectorStoreRecordKeyPropertyModel> KeyProperties { get; }
+
+    /// <summary>
+    /// The dataproperties of the record.
+    /// </summary>
+    public IReadOnlyList<VectorStoreRecordDataPropertyModel> DataProperties { get; }
+
+    /// <summary>
+    /// The vector properties of the record.
+    /// </summary>
+    public IReadOnlyList<VectorStoreRecordVectorPropertyModel> VectorProperties { get; }
+
+    /// <summary>
+    /// All properties of the record, of all types.
+    /// </summary>
+    public IReadOnlyList<VectorStoreRecordPropertyModel> Properties { get; }
+
+    /// <summary>
+    /// All properties of the record, of all types, indexed by their model name.
+    /// </summary>
+    public IReadOnlyDictionary<string, VectorStoreRecordPropertyModel> PropertyMap { get; }
+
+    internal VectorStoreRecordModel(
+        Type recordClrType,
+        IReadOnlyList<VectorStoreRecordKeyPropertyModel> keyProperties,
+        IReadOnlyList<VectorStoreRecordDataPropertyModel> dataProperties,
+        IReadOnlyList<VectorStoreRecordVectorPropertyModel> vectorProperties,
+        IReadOnlyDictionary<string, VectorStoreRecordPropertyModel> propertyMap)
+    {
+        this._recordClrType = recordClrType;
+        this.KeyProperties = keyProperties;
+        this.DataProperties = dataProperties;
+        this.VectorProperties = vectorProperties;
+        this.PropertyMap = propertyMap;
+        this.Properties = propertyMap.Values.ToList();
+    }
+
+    /// <summary>
+    /// Returns the single key property in the model, and throws if there are multiple key properties.
+    /// Suitable for connectors where validation is in place for single keys only (<see cref="VectorStoreRecordModelBuildingOptions.SupportsMultipleKeys"/>).
+    /// </summary>
+    public VectorStoreRecordKeyPropertyModel KeyProperty => this._singleKeyProperty ??= this.KeyProperties.Single();
+
+    /// <summary>
+    /// Returns the single vector property in the model, and throws if there are multiple vector properties.
+    /// Suitable for connectors where validation is in place for single vectors only (<see cref="VectorStoreRecordModelBuildingOptions.SupportsMultipleVectors"/>).
+    /// </summary>
+    public VectorStoreRecordVectorPropertyModel VectorProperty => this._singleVectorProperty ??= this.VectorProperties.Single();
+
+    /// <summary>
+    /// Instantiates a new record of the specified type.
+    /// </summary>
+    // TODO: the pattern of first instantiating via parameterless constructor and then populating the properties isn't compatible
+    // with read-only types, where properties have no setters. Supporting those would be problematic given the that different
+    // connectors have completely different representations of the data coming back from the database, and which needs to be
+    // populated.
+    public TRecord InstantiateRecord<TRecord>()
```

Comment:
How about just `CreateRecord` ?

```suggestion
    public TRecord CreateRecord<TRecord>()
```